### PR TITLE
Fix crash due to ambiguous cache file paths

### PR DESCRIPTION
**Pull request by fangly**
Here is a fix for #212. It takes the approach described in the same ticket, i.e. give different packages a different cache directory, give each file its path relative to the package. This ensures that no two R source files end up having the same cache location.

One thing to note is that `find_package()` and `pkg_name()` are called frequently in this fix. I am not sure of the performance penalty (e.g. `pkg_name()` needs to read the DESCRIPTION file each time). It may be worth somehow saving the package name and location once they have been determined.

### CODE REVIEW

**Comment by jimhester** [comment]
This solution is too complicated, you have the package root and the file path relative to the package root, you just need to append them and use that as the cache path.

Also I see no reason to rename the argument, we need to maximize backwards compatibility unless there is substantive improvement.

**Comment by fangly** [comment]
Ok, I renamed the argument in the two functions, though I do think that its name is not precise enough and its use particularly confusing in `clear_cache(file, path)`.

I partially disagree with your statement about my solution being to complicated:

1. As far as I can tell, there is no guarantee that the filenames passed to `lint()` are relative paths. With `lint_package(".", relative_path = TRUE)`, I get absolute paths (on the master branch) `lint_package()` calls `normalizePath(files)`, which is the correct thing to do in my opinion. Calling `lint()` directly, one may get relative paths, but they are relative to the current directory and not necessarily the package directory.

2. An alternative solution (that could be considered simpler) is to create a hash on the absolute file path.

**Comment by fangly** [comment]
Hi Jim,
I did implement the alternative solution mentionned above. Please do have a look at the code again. 
Going one little step farther, one could also merge `get_cache_file_path()` into `digest_content()`. Note that the last commit is not really related to caching.
Cheers,
Florent

File: R/lint.R [hunk 1]
```diff
@@ -132,16 +133,19 @@ reorder_lints <- function(lints) {
...
 #' @export
 lint_package <- function(path = ".", relative_path = TRUE, ...) {
   path <- find_package(path)
 
   read_settings(path)
   on.exit(clear_settings, add = TRUE)
 
-  names(settings$exclusions) <- normalizePath(file.path(path, names(settings$exclusions)))
-  exclusions = force(settings$exclusions)
+  args <- list(...)
```
Comment (jimhester) [inline R/lint.R]: We should add `exclusions` to the argument list of `lint_package()` after the `...` and `#' @includeParams lint`, then modify the exclusions as needed and just use a standard call to `lint()` as before.

**Comment by fangly** [comment]
Very well, this is now done.

**Comment by fangly** [comment]
Please, remember to merge this branch, Jim.

**Comment by jimhester** [comment]
Thanks again for working on this!

**Pull request merged**